src/platform/riot/uart_channel.c

```diff
@@ -0,0 +1,161 @@
+    }
+  }
+}
+
+void UARTSyncChannel_poll(NetworkChannel *untyped_self) {
+  UARTSyncChannel *self = (UARTSyncChannel *)untyped_self;
+
+  int bytes_left = deserialize_from_protobuf(&self->output, self->read_buffer, self->read_index);
+  UART_CHANNEL_DEBUG("Bytes Left after attempted to deserialize %d", bytes_left);
+
+  if (bytes_left >= 0) {
+    int read_index = self->read_index;
+    self->read_index = bytes_left;
+
+    memcpy(self->read_buffer, self->read_buffer + (read_index - bytes_left), bytes_left);
+
+    if (self->receive_callback != NULL) {
+      UART_CHANNEL_DEBUG("calling user callback!");
+      self->receive_callback(self->federated_connection, &self->output);
+    }
+  }
+}
+
+void *_UARTAsyncChannel_decode_loop(void *arg) {
+  UARTAsyncChannel *self = (UARTAsyncChannel *)arg;
+  mutex_lock(&self->receive_lock);
+
+  UART_CHANNEL_DEBUG("Entering decoding loop");
+  while (true) {
+    cond_wait(&self->receive_cv, &self->receive_lock);
+
+    UARTSyncChannel_poll((NetworkChannel *)&self->super);
+
+    thread_yield_higher();
+  }
+
+  return NULL;
+}
+
+void UARTSyncChannel_ctor(UARTSyncChannel *self, Environment *env, uint32_t uart_device, uint32_t baud) {
+  assert(self != NULL);
+  assert(env != NULL);
+
+  self->uart_dev = UART_DEV(uart_device);
+
+  int result = uart_init(self->uart_dev, baud, _UARTSyncChannel_receive_callback, self);
+
+  if (result == -ENODEV) {
+    UART_CHANNEL_ERR("Invalid UART device!");
+  } else if (result == ENOTSUP) {
+    UART_CHANNEL_ERR("Given configuration to uart is not supported!");
+  } else if (result < 0) {
+    UART_CHANNEL_ERR("UART Init error occurred");
+  }
+
+  result = uart_mode(self->uart_dev, UART_DATA_BITS_8, UART_PARITY_EVEN, UART_STOP_BITS_2);
+
+  if (result != UART_OK) {
+    UART_CHANNEL_ERR("Problem to configure UART device!");
+    return;
+  }
+
+  self->super.super.category = NETWORK_CHANNEL_CATEGORY_SYNC;
+  self->super.super.expected_connect_duration = UART_CHANNEL_EXPECTED_CONNECT_DURATION;
+  self->super.super.type = NETWORK_CHANNEL_TYPE_UART;
+  self->super.super.is_connected = UARTSyncChannel_is_connected;
+  self->super.super.open_connection = UARTSyncChannel_open_connection;
+  self->super.super.close_connection = UARTSyncChannel_close_connection;
+  self->super.super.send_blocking = UARTSyncChannel_send_blocking;
+  self->super.super.register_receive_callback = UARTSyncChannel_register_receive_callback;
+  self->super.super.free = UARTSyncChannel_free;
+  self->super.poll = UARTSyncChannel_poll;
+
+  // Concrete fields
+  self->read_index = 0;
+  self->receive_callback = NULL;
+  self->federated_connection = NULL;
+  self->state = NETWORK_CHANNEL_STATE_CONNECTED;
+}
+
+void UARTAsyncChannel_ctor(UARTAsyncChannel *self, Environment *env, uint32_t uart_device, uint32_t baud) {
+  UARTSyncChannel_ctor(&self->super, env, uart_device, baud);
+
+  cond_init(&self->receive_cv);
+  mutex_init(&self->receive_lock);
+
+  // Create decoding thread
```

Comment:
Both "connection thread" and "decode thread" is used to describe this thread and associated variables. Wuold be good to stick to one

Reply:
renamed it to `decoding_thread`